CRM/Member/BAO/Membership.php

```diff
@@ -344,22 +344,11 @@ public static function create(&$params, $ids = []) {
 
     // Record contribution for this membership and create a MembershipPayment
     // @todo deprecate this.
-    if (!empty($params['contribution_status_id']) && empty($params['relate_contribution_id'])) {
+    if (!empty($params['contribution_status_id'])) {
       $memInfo = array_merge($params, ['membership_id' => $membership->id]);
       $params['contribution'] = self::recordMembershipContribution($memInfo);
     }
 
-    // Add/update MembershipPayment record for this membership if it is a related contribution
-    // @todo remove this - called from one remaining place in CRM_Member_Form_Membership
-    if (!empty($params['relate_contribution_id'])) {
```

Comment:
Everything to do with 'relate_contribution_id' is the weird magic that made it work despite doing things all out of order - getting rid of this sticky tape is 'the goal'